File: Sorting/23056.js
const fs = require('fs');
const [nm, ...input] = fs.readFileSync('/dev/stdin', 'utf8').trim().split('\n');
const [n, m] = nm.trim().split(' ').map(Number);

let students = [];
let cnt = Array(n + 1).fill(0);
for(let i = 0; i < input.length; i++){
	let [clss, name] = input[i].trim().split(' ');
	let cls = Number(clss);
	if(cls === 0) break;
	let team; // 청팀: 1, 백팀: 2
	if(cls % 2 === 0) team = 2;
	else team = 1;
	
	let student = {
		team: team,
		cls: cls,
		name: name
	};
	if(cnt[cls] < m){
		students.push(student);
		cnt[cls]++;
	}
}
students.sort((a, b) => {
	if(a.cls === b.cls) {
		if(a.name.length == b.name.length) return a.name - b.name;
		else return a.name.length - b.name.length;
	}
	if(a.team === b.team) return a.cls - b.cls;
	return a.team - b.team;
});

for(let i = 0; i < students.length; i++){
 	console.log(students[i].cls + ' ' + students[i].name);
}

File: Sorting/23056.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

struct Student{
	int team; // 청팀: 1, 백팀: 2
	int clas;
	string name;
};

bool cmp(Student a, Student b){
	if(a.clas == b.clas){
		if(a.name.length() == b.name.length()) return a.name < b.name;
		return a.name.length() < b.name.length();
	}
	if(a.team == b.team) return a.clas < b.clas;
	return a.team < b.team;
}

int main() {
	int n, m;
	cin >> n >> m;
	vector<Student> students;
	int cnt[11] = {0, };
	while(true){
		int cls; string nme;
		cin >> cls >> nme;
		if(cls == 0) break;
		
		int tm;
		if(cls % 2 == 0) tm = 2;
		else tm = 1;
		if(cnt[cls] < m) {
			students.push_back({tm, cls, nme});
			cnt[cls]++;
		}
	}
	sort(students.begin(), students.end(), cmp);
	for(int i = 0; i < students.size(); i++){
		cout << students[i].clas << " " << students[i].name << '\n';
	}
	return 0;
}
